dsls/sql/src/main/java/org/apache/beam/dsls/sql/BeamSql.java

```diff
@@ -0,0 +1,292 @@
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.impl.ScalarFunctionImpl;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.calcite.tools.Frameworks;
+import org.apache.calcite.tools.RelConversionException;
+import org.apache.calcite.tools.ValidationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@code BeamSql} is the DSL interface of BeamSQL. It translates a SQL query as a
+ * {@link PTransform}, so developers can use standard SQL queries in a Beam pipeline.
+ *
+ * <h1>Beam SQL DSL usage:</h1>
+ * A typical pipeline with Beam SQL DSL is:
+ * <pre>
+ *{@code
+PipelineOptions options = PipelineOptionsFactory.create();
+Pipeline p = Pipeline.create(options);
+
+//create table from TextIO;
+TableSchema tableASchema = ...;
+PCollection<BeamSqlRow> inputTableA = p.apply(TextIO.read().from("/my/input/patha"))
+    .apply(BeamSql.fromTextRow(tableASchema));
+TableSchema tableBSchema = ...;
+PCollection<BeamSqlRow> inputTableB = p.apply(TextIO.read().from("/my/input/pathb"))
+    .apply(BeamSql.fromTextRow(tableBSchema));
+
+//run a simple query, and register the output as a table in BeamSql;
+String sql1 = "select MY_FUNC(c1), c2 from TABLE_A";
+PCollection<BeamSqlRow> outputTableA = inputTableA.apply(BeamSql.simpleQuery(sql1))
+        .withUdf("MY_FUNC", myFunc);
+
+//run a JOIN with one table from TextIO, and one table from another query
+PCollection<BeamSqlRow> outputTableB = PCollectionTuple.of(
+    new TupleTag<BeamSqlRow>("TABLE_O_A"), outputTableA)
+                .and(new TupleTag<BeamSqlRow>("TABLE_B"), inputTableB)
+    .apply(BeamSql.query("select * from TABLE_O_A JOIN TABLE_B where ..."));
+
+//output the final result with TextIO
+outputTableB.apply(BeamSql.toTextRow()).apply(TextIO.write().to("/my/output/path"));
+
+p.run().waitUntilFinish();
+ * }
+ * </pre>
+ *
+ * <h1>Beam SQL DSL vs CLI client</h1>
+ * Beam SQL CLI client(pending) is another way to leverage SQL to process data,
+ * which is built on DSL interfaces.
+ *
+ */
+public class BeamSql implements Serializable {
+  private static final Logger LOG = LoggerFactory.getLogger(BeamSql.class);
+
+  private static SchemaPlus schema;
+  private static BeamQueryPlanner planner;
+
+  static{
+    //Disable Assertion in Calcite temporally, it would not ignore any grammar errors.
+    ClassLoader.getSystemClassLoader().setDefaultAssertionStatus(false);
+    schema = Frameworks.createRootSchema(true);
+    planner = new BeamQueryPlanner(schema);
+  }
+
+  /**
+   * Add a UDF function.
+   */
+  public static void addUDFFunction(String functionName, Class<?> className, String methodName){
```

Comment:
I would expose it as a global method, `create method` in CLI would need this function.